### PR TITLE
Move to use async functions

### PR DESCRIPTION
**Pull request by hueniverse**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -84,25 +65,23 @@ const options = {
...
-
-    // buffer the response stream
-    Wreck.read(res, null, (err, body) => {
-        /* do stuff */
-    });
-};
-
-const req = wreck.request(method, uri, options, optionalCallback);
+const promise = wreck.request(method, uri, options);
+// Use promise.req.abort() to terminate the request early
```
Comment (kanongil) [inline README.md]: This is not that simple. `promise.req` will be the wrong one, if the request has been redirected.

File: test/index.js [hunk 2]
```diff
@@ -1356,142 +736,86 @@ describe('request()', () => {
...
-                done();
-            });
-        });
+        const server = await internals.server(handler);
+        try {
+            await Wreck.request('get', 'http://localhost:' + server.address().port, { timeout: 100 });
+        }
+        catch (err) {
+            expect(err.output.statusCode).to.equal(504);
+        }
```
Comment (kanongil) [inline test/index.js]: This is not a safe way to test. The `expect()` won't be called if the request is successful, causing it to unexpectedly pass.

We probably need something like https://github.com/hapijs/code/issues/106, for a simple safe solution.

File: README.md (hunk 1 above)
Comment (hueniverse) [reply README.md]: How is this different from previous code?

Comment (kanongil) [reply README.md]: Aborting is useful for some use-cases. Anyway, it's primarily the comment I have an issue with, as it implies that aborting will just work like that.

Comment (hueniverse) [reply README.md]: Fair.

File: README.md [hunk 3]
```diff
@@ -84,25 +65,26 @@ const options = {
...
+}
+```
 
-    // buffer the response stream
-    Wreck.read(res, null, (err, body) => {
-        /* do stuff */
-    });
-};
+Use `promise.req.abort()` to terminate the request early. Note that this is limited to the initial request only.
+If the was request already redirected, aborting the original request will not abort execution of pending redirections.
```
Comment (geek) [inline README.md]: missing a word: 
>If the was

File: README.md [hunk 4]
```diff
@@ -168,115 +145,112 @@ Returns an instance of the node.js [ClientRequest](http://nodejs.org/api/http.ht
...
+
+Returns a promise that resolves into an object with the following properties:
+    - `res` - The [HTTP Incoming Message](https://nodejs.org/api/http.html#http_class_http_incomingmessage)
        object, which is a readable stream that has "ended" and contains no more data to read.
     - `payload` - The payload in the form of a Buffer or (optionally) parsed JavaScript object (JSON).
 
-If a callback function is provided then an instance of the node.js [ClientRequest](http://nodejs.org/api/http.html#http_class_http_clientrequest) object is returned.
-If no callback function is provided then a Promise is returned that resolves an object with the following structure: `{ req, res, payload }`;
-
+Thows any error that may have occurred during handling of the request or a Boom error object if the response has an error status
```
Comment (geek) [inline README.md]: typo: 
> Thows

File: README.md [hunk 5]
```diff
@@ -168,115 +145,112 @@ Returns an instance of the node.js [ClientRequest](http://nodejs.org/api/http.ht
...
-    - `response` - The [HTTP Incoming Message](https://nodejs.org/api/http.html#http_class_http_incomingmessage)
+
+Returns a promise that resolves into an object with the following properties:
+    - `res` - The [HTTP Incoming Message](https://nodejs.org/api/http.html#http_class_http_incomingmessage)
        object, which is a readable stream that has "ended" and contains no more data to read.
     - `payload` - The payload in the form of a Buffer or (optionally) parsed JavaScript object (JSON).
 
-If a callback function is provided then an instance of the node.js [ClientRequest](http://nodejs.org/api/http.html#http_class_http_clientrequest) object is returned.
-If no callback function is provided then a Promise is returned that resolves an object with the following structure: `{ req, res, payload }`;
+Thows any error that may have occurred during handling of the request or a Boom error object if the response has an error status
```
Comment (geek) [inline README.md]: typo:
>Thows

File: README.md [hunk 6]
```diff
@@ -168,115 +145,112 @@ Returns an instance of the node.js [ClientRequest](http://nodejs.org/api/http.ht
...
+
+Returns a promise that resolves into an object with the following properties:
+    - `res` - The [HTTP Incoming Message](https://nodejs.org/api/http.html#http_class_http_incomingmessage)
        object, which is a readable stream that has "ended" and contains no more data to read.
     - `payload` - The payload in the form of a Buffer or (optionally) parsed JavaScript object (JSON).
 
-If a callback function is provided then an instance of the node.js [ClientRequest](http://nodejs.org/api/http.html#http_class_http_clientrequest) object is returned.
-If no callback function is provided then a Promise is returned that resolves an object with the following structure: `{ req, res, payload }`;
-
+Thows any error that may have occurred during handling of the request or a Boom error object if the response has an error status
```
Comment (geek) [inline README.md]: typo:
>Thows

File: README.md [hunk 7]
```diff
@@ -168,115 +145,112 @@ Returns an instance of the node.js [ClientRequest](http://nodejs.org/api/http.ht
...
```
Comment (geek) [inline README.md]: typo:
>Thows

File: README.md [hunk 8]
```diff
@@ -168,115 +145,112 @@ Returns an instance of the node.js [ClientRequest](http://nodejs.org/api/http.ht
...
```
Comment (geek) [inline README.md]: typo:
>Thows

File: test/index.js [hunk 9]
```diff
@@ -32,567 +31,202 @@ const internals = {
...
-    it('requests a resource with callback', (done) => {
-
-        const server = Http.createServer((req, res) => {
-
-            res.writeHead(200, { 'Content-Type': 'text/plain' });
-            res.end(internals.payload);
-        });
-
-        server.listen(0, () => {
+    it('requests a resource with callback', async () => {
```
Comment (geek) [inline test/index.js]: Can update the test title: `retrieves a resource`

**Pull request merged**